src/main/java/com/puppycrawl/tools/checkstyle/checks/design/OneTopLevelClassCheck.java

```diff
@@ -126,41 +113,49 @@
     @Override
     public void beginTree(DetailAST rootAST) {
         publicTypeFound = false;
-        lineNumberTypeMap.clear();
 
         DetailAST currentNode = rootAST;
         while (currentNode != null) {
-            if (currentNode.getType() == TokenTypes.CLASS_DEF
-                    || currentNode.getType() == TokenTypes.ENUM_DEF
-                    || currentNode.getType() == TokenTypes.INTERFACE_DEF) {
-                if (isPublic(currentNode)) {
-                    publicTypeFound = true;
-                }
-                else {
-                    final String typeName = currentNode
-                            .findFirstToken(TokenTypes.IDENT).getText();
-                    lineNumberTypeMap.put(currentNode, typeName);
-                }
+            if (isTypeDef(currentNode) && isPublic(currentNode)) {
+                publicTypeFound = true;
```

Comment:
Now the code is trimmed down to a bare 2 methods, both look like they pretty much do the same thing. `beginTree` only checks for a public type and `finishTree` does the final violations. Other than that they both do the same thing and scan all the siblings connected to the root.

I question if we should just merge the 2 methods together and remove all the fields, and see if we can do the AST sibling scan only once. What is the benefit now of keeping them in different methods?

Reply:
I've managed to squash everything into one pass, so now there's only one method and no instance fields.

The only side effect is that our first type is stored somewhere and we only check if it is a violation at the end of the loop. The UTs aren't affected, so it should be all good. Might do another regression report later.

Reply:
New regression is required.

Reply:
Updated.

New: https://wltan.github.io/checkstyle-reports/2020-03-26/onetoplevel-detailast-regression/diff/
Old: https://wltan.github.io/checkstyle-reports/2020-03-19/onetoplevel-detailast-regression/diff/